### PR TITLE
Enhance openflow communication performance

### PR DESCRIPTION
**Pull request by lfu-ps**
1. "include/libfluid-*" and "src/libfluid-*" are copied from https://github.com/OpenNetworkingFoundation/libfluid_base and msg, they are third party open source libraries so you can skip reviewing, btw it is Apache 2.0 license (https://github.com/OpenNetworkingFoundation/libfluid_base/blob/master/LICENSE)
2. please focus on how we leverage the libfluid fundamental, meaning the derived classes from it
3. switch to openvswitch release version dependencies (https://www.openvswitch.org/download/) instead of the git src code version
4. migrate ovs control functions to be compatible with the ovs release headers and lib
5. style enforcement can be added on later, not fully completed on every header and src file

### CODE REVIEW

**Comment by lfu-ps** [comment]
[https://github.com/futurewei-cloud/alcor-control-agent/issues/248](url)
Due to the new mechanism, the issue above is fixed with the refactoring. Tested with 1000 ports successfully.

File: build/test.sh [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+#
+#     The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+#
+#     THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+#     FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+#     WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+#!/bin/bash
+
+BUILD="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
```
Comment (cj-chung) [inline build/test.sh:17]: @lfu-ps What this test.sh for? It seems very similar to `aca-machine-init.sh`. Should we rename the file more descriptive or combine `test.sh` and `aca-machine-init.sh` into one and give some options to it for the diff part.

File: src/aca_main.cpp [hunk 2]
```diff
@@ -263,17 +285,39 @@ int main(int argc, char *argv[])
...
+
+  // set bridge controller will clean up flows
+  aca_ovs_l2_programmer::ACA_OVS_L2_Programmer::get_instance().setup_ovs_controller(g_ovs_ctrl_address, g_ovs_ctrl_port);
+
+  // start local ovs server (openflow controller)
+  g_ovs_ctrl = new OFController(switch_dpid_map, port_id_map, g_ovs_ctrl_address.c_str(), g_ovs_ctrl_port);
+  g_ovs_ctrl->start();
+
+  // pass ovs_ctrl handle to l2 programmer
+  aca_ovs_l2_programmer::ACA_OVS_L2_Programmer::get_instance().set_openflow_controller(g_ovs_ctrl);
```
Comment (cj-chung) [inline src/aca_main.cpp:305]: @lfu-ps It seems like we can put all these statements in a dedicated function (such as `setup_openflow` in `ACA_OVS_L2_Programmer`, then we can make only 1 call here.

File: include/ovs_control.h [hunk 3]
```diff
@@ -12,144 +12,157 @@
...
+//#include <openvswitch/ofp-port.h>
+//#include <openvswitch/ofp-table.h>
+//#include <openvswitch/ofp-protocol.h>
+//#include <openvswitch/ofp-switch.h>
+//#include <openvswitch/ofp-flow.h>
 
-extern "C" { 
-    struct unixctl_conn;
+extern "C" {
+struct unixctl_conn;
```
Comment (cj-chung) [inline include/ovs_control.h:37]: @lfu-ps If we don't use `vconn`, we probably don't need this struct too.

Comment (lfu-ps) [reply include/ovs_control.h:37]: I'm testing removing the entire ovs_control.h/cpp, but currently keeping it for function reference (even it's not invoked).

File: src/aca_main.cpp (hunk 2 above)
Comment (lfu-ps) [reply src/aca_main.cpp:305]: Fixed and resolved.

File: build/Dockerfile [hunk 4]
```diff
@@ -157,6 +157,11 @@ ENV OPENFLOW_HEADERS='include/openflow/intel-ext.h \
     include/openflow/openflow-common.h \
     include/openflow/openflow.h '
 RUN echo "5--- installing openvswitch dependancies ---" && \
+    rm -f /tmp/get-pip.py > /dev/null 2>&1 && \
+    apt-get install -y python2.7 && \
+    wget https://bootstrap.pypa.io/pip/2.7/get-pip.py -O /tmp/get-pip.py && \
+    python2.7 /tmp/get-pip.py && \
+    pip2 install six && \
```
Comment (pkommoju) [inline build/Dockerfile:164]: I kept getting permission denied error, got tired of it and tacked sudo infron of all of them.

Comment (lfu-ps) [reply build/Dockerfile]: Added in next commit

File: build/aca-machine-init.sh [hunk 5]
```diff
@@ -172,6 +172,11 @@ OPENFLOW_HEADERS="include/openflow/intel-ext.h \
     include/openflow/openflow-common.h \
     include/openflow/openflow.h "
 echo "6--- installing openvswitch dependancies ---" && \
+    rm -f /tmp/get-pip.py > /dev/null 2>&1 && \
+    apt-get install -y python2.7 && \
+    wget https://bootstrap.pypa.io/pip/2.7/get-pip.py -O /tmp/get-pip.py && \
+    python2.7 /tmp/get-pip.py && \
+    pip2 install six && \
     apt-get install -y libevent-dev && \
```
[the remark this answers is not in the document]
Comment (lfu-ps) [reply build/aca-machine-init.sh:180]: Next commit does it

**Review by cj-chung — APPROVED** [review]
LGTM

**Review by xieus — COMMENTED** [review]
A few minor comments.

File: include/libfluid-base/OFClient.hh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+namespace fluid_base {
+
+class OFClient : private BaseOFClient, private OFConnectionProcessor, public OFHandler {
+public:
+    OFClient(
+            const std::string& addr,
+            const bool domainsocket,
+            const int port,
+            const bool secure,
```
[the remark this answers is not in the document]
Comment (lfu-ps) [reply include/libfluid-base/OFClient.hh:18]: We assume, and using secure=false in current perf report. Need to test secure true (means TLS).

File: include/libfluid-base/OFClient.hh [hunk 7]
```diff
@@ -0,0 +1,41 @@
+#pragma once
+
```
[the remark this answers is not in the document]
Comment (lfu-ps) [reply include/libfluid-base/OFClient.hh:2]: Resolved

File: include/libfluid-base/OFClient.hh [hunk 8]
```diff
@@ -1,3 +1,18 @@
+// Copyright (c) 2014 Open Networking Foundation
+// Copyright 2021 The Alcor Authors
```
Comment (xieus) [inline include/libfluid-base/OFClient.hh:2]: We could use "Copyright 2020 Futurewei Cloud" throughout (previously we did use "Alcor Authors" but changed earlier this year).

Comment (lfu-ps) [reply include/libfluid-base/OFClient.hh]: Renamed

File: src/aca_main.cpp [hunk 9]
```diff
@@ -71,7 +80,7 @@ std::atomic_ulong g_total_vpcs_table_mutex_time(0);
 std::atomic_ulong g_total_update_GS_time(0);
 
 bool g_demo_mode = false;
-bool g_debug_mode = false;
+bool g_debug_mode = true;
```
Comment (xieus) [inline src/aca_main.cpp:83]: Are we going to keep debug mode true?

**Review by xieus — APPROVED** [review]
Thank you @lfu-ps. This PR optimizes the comm layer significantly.

File: src/aca_main.cpp (hunk 9 above)
Comment (lfu-ps) [reply src/aca_main.cpp:83]: No, it was only for printing debug logs. Reverted. Good catch, thanks.

**Pull request merged**